File: src/models/PhotoItem.m
/**
 * PhotoItem.m
 * Photo model implementation with metadata loading
 */

#import "PhotoItem.h"
#import <ImageIO/ImageIO.h>
#import "../utils/DateFormatters.h"

@interface PhotoItem ()

@property (nonatomic, copy, readwrite) NSString *path;
@property (nonatomic, copy, readwrite) NSString *name;
@property (nonatomic, assign, readwrite) uint64_t fileSize;
@property (nonatomic, strong, readwrite) NSDate *createdDate;
@property (nonatomic, strong, readwrite) NSDate *modifiedDate;

@property (nonatomic, assign, readwrite) CGSize dimensions;
@property (nonatomic, copy, readwrite, nullable) NSString *cameraMake;
@property (nonatomic, copy, readwrite, nullable) NSString *cameraModel;
@property (nonatomic, copy, readwrite, nullable) NSString *exposureTime;
@property (nonatomic, copy, readwrite, nullable) NSString *fNumber;
@property (nonatomic, copy, readwrite, nullable) NSNumber *iso;
@property (nonatomic, copy, readwrite, nullable) NSString *focalLength;
@property (nonatomic, copy, readwrite, nullable) NSDate *dateTaken;
@property (nonatomic, copy, readwrite, nullable) NSString *colorSpace;
@property (nonatomic, assign, readwrite) BOOL metadataLoaded;

@end

@implementation PhotoItem

- (instancetype)initWithPath:(NSString *)path
                        name:(NSString *)name
                    fileSize:(uint64_t)fileSize
                 createdDate:(NSDate *)createdDate
                modifiedDate:(NSDate *)modifiedDate {
    self = [super init];
    if (self) {
        _path = [path copy];
        _name = [name copy];
        _fileSize = fileSize;
        _createdDate = createdDate;
        _modifiedDate = modifiedDate;
        _dimensions = CGSizeZero;
        _metadataLoaded = NO;
    }
    return self;
}

- (void)loadMetadata {
    if (self.metadataLoaded) {
        return;
    }
    
    NSURL *url = [NSURL fileURLWithPath:self.path];
    CGImageSourceRef source = CGImageSourceCreateWithURL((__bridge CFURLRef)url, NULL);
    
    if (source == NULL) {
        self.metadataLoaded = YES;
        return;
    }
    
    NSDictionary *properties = (__bridge_transfer NSDictionary *)CGImageSourceCopyPropertiesAtIndex(source, 0, NULL);
    
    if (properties) {
        /* Dimensions */
        NSNumber *width = properties[(__bridge NSString *)kCGImagePropertyPixelWidth];
        NSNumber *height = properties[(__bridge NSString *)kCGImagePropertyPixelHeight];
        if (width && height) {
            self.dimensions = CGSizeMake(width.doubleValue, height.doubleValue);
        }
        
        /* Color space */
        self.colorSpace = properties[(__bridge NSString *)kCGImagePropertyColorModel];
        
        /* EXIF data */
        NSDictionary *exif = properties[(__bridge NSString *)kCGImagePropertyExifDictionary];
        if (exif) {
            self.exposureTime = [self formatExposureTime:exif[(__bridge NSString *)kCGImagePropertyExifExposureTime]];
            self.fNumber = [self formatFNumber:exif[(__bridge NSString *)kCGImagePropertyExifFNumber]];
            self.iso = [exif[(__bridge NSString *)kCGImagePropertyExifISOSpeedRatings] firstObject];
            self.focalLength = [self formatFocalLength:exif[(__bridge NSString *)kCGImagePropertyExifFocalLength]];
            self.dateTaken = [self parseExifDate:exif[(__bridge NSString *)kCGImagePropertyExifDateTimeOriginal]];
        }
        
        /* TIFF data (camera info) */
        NSDictionary *tiff = properties[(__bridge NSString *)kCGImagePropertyTIFFDictionary];
        if (tiff) {
            self.cameraMake = tiff[(__bridge NSString *)kCGImagePropertyTIFFMake];
            self.cameraModel = tiff[(__bridge NSString *)kCGImagePropertyTIFFModel];
        }
    }
    
    CFRelease(source);
    self.metadataLoaded = YES;
}

#pragma mark - Formatting Helpers

- (NSString *)formatExposureTime:(NSNumber *)exposure {
    if (!exposure) {
        return nil;
    }
    
    double value = exposure.doubleValue;
    if (value >= 1.0) {
        return [NSString stringWithFormat:@"%.1fs", value];
    } else if (value > 0) {
        return [NSString stringWithFormat:@"1/%.0fs", 1.0 / value];
    }
    return nil;
}

- (NSString *)formatFNumber:(NSNumber *)fNumber {
    if (!fNumber) {
        return nil;
    }
    return [NSString stringWithFormat:@"f/%.1f", fNumber.doubleValue];
}

- (NSString *)formatFocalLength:(NSNumber *)focal {
    if (!focal) {
        return nil;
    }
    return [NSString stringWithFormat:@"%.0fmm", focal.doubleValue];
}

- (NSDate *)parseExifDate:(NSString *)dateString {
    return [DateFormatters dateFromExifString:dateString];
}

+ (NSString *)formattedFileSize:(uint64_t)bytes {
    static const char *units[] = {"B", "KB", "MB", "GB", "TB"};
    int unitIndex = 0;
    double size = (double)bytes;
    
    while (size >= 1024.0 && unitIndex < 4) {
        size /= 1024.0;
        unitIndex++;
    }
    
    if (unitIndex == 0) {
        return [NSString stringWithFormat:@"%llu %s", bytes, units[unitIndex]];
    }
    return [NSString stringWithFormat:@"%.1f %s", size, units[unitIndex]];
}

@end

File: src/swift/Models/PhotoItem.swift
/**
 * PhotoItem.swift
 * Model representing a single photo with metadata
 */

import Foundation
import AppKit
import ImageIO

struct PhotoItem: Identifiable, Hashable {
    let id: UUID
    let path: String
    let name: String
    let fileSize: UInt64
    let createdDate: Date
    let modifiedDate: Date
    let index: Int
    
    // Lazy-loaded metadata
    private(set) var dimensions: CGSize = .zero
    private(set) var cameraMake: String?
    private(set) var cameraModel: String?
    private(set) var exposureTime: String?
    private(set) var fNumber: String?
    private(set) var iso: Int?
    private(set) var focalLength: String?
    private(set) var dateTaken: Date?
    private(set) var colorSpace: String?
    private(set) var metadataLoaded: Bool = false
    
    init(path: String, name: String, fileSize: UInt64, createdDate: Date, modifiedDate: Date, index: Int) {
        self.id = UUID()
        self.path = path
        self.name = name
        self.fileSize = fileSize
        self.createdDate = createdDate
        self.modifiedDate = modifiedDate
        self.index = index
    }
    
    mutating func loadMetadata() {
        guard !metadataLoaded else { return }
        
        let url = URL(fileURLWithPath: path)
        guard let source = CGImageSourceCreateWithURL(url as CFURL, nil) else {
            metadataLoaded = true
            return
        }
        
        guard let properties = CGImageSourceCopyPropertiesAtIndex(source, 0, nil) as? [String: Any] else {
            metadataLoaded = true
            return
        }
        
        // Dimensions
        if let width = properties[kCGImagePropertyPixelWidth as String] as? Double,
           let height = properties[kCGImagePropertyPixelHeight as String] as? Double {
            dimensions = CGSize(width: width, height: height)
        }
        
        // Color space
        colorSpace = properties[kCGImagePropertyColorModel as String] as? String
        
        // EXIF data
        if let exif = properties[kCGImagePropertyExifDictionary as String] as? [String: Any] {
            if let exposure = exif[kCGImagePropertyExifExposureTime as String] as? Double {
                exposureTime = Self.formatExposureTime(exposure)
            }
            if let fNum = exif[kCGImagePropertyExifFNumber as String] as? Double {
                fNumber = String(format: "f/%.1f", fNum)
            }
            if let isoArray = exif[kCGImagePropertyExifISOSpeedRatings as String] as? [Int], 
               let firstISO = isoArray.first {
                iso = firstISO
            }
            if let focal = exif[kCGImagePropertyExifFocalLength as String] as? Double {
                focalLength = String(format: "%.0fmm", focal)
            }
            if let dateString = exif[kCGImagePropertyExifDateTimeOriginal as String] as? String {
                dateTaken = DateFormatters.parseExifDate(dateString)
            }
        }
        
        // TIFF data (camera info)
        if let tiff = properties[kCGImagePropertyTIFFDictionary as String] as? [String: Any] {
            cameraMake = tiff[kCGImagePropertyTIFFMake as String] as? String
            cameraModel = tiff[kCGImagePropertyTIFFModel as String] as? String
        }
        
        metadataLoaded = true
    }
    
    private static func formatExposureTime(_ value: Double) -> String {
        if value >= 1.0 {
            return String(format: "%.1fs", value)
        } else if value > 0 {
            return String(format: "1/%.0fs", 1.0 / value)
        }
        return ""
    }
    
    static func formattedFileSize(_ bytes: UInt64) -> String {
        let units = ["B", "KB", "MB", "GB", "TB"]
        var size = Double(bytes)
        var unitIndex = 0
        
        while size >= 1024 && unitIndex < units.count - 1 {
            size /= 1024
            unitIndex += 1
        }
        
        if unitIndex == 0 {
            return "\(bytes) \(units[unitIndex])"
        }
        return String(format: "%.1f %@", size, units[unitIndex])
    }
    
    func hash(into hasher: inout Hasher) {
        hasher.combine(path)
    }
    
    static func == (lhs: PhotoItem, rhs: PhotoItem) -> Bool {
        lhs.path == rhs.path
    }
}
